### PR TITLE
[WIP] Feature/type search

### PR DESCRIPTION
**Pull request by klaftertief**
This adds a separate search page, inspired by [Hoogle](https://www.haskell.org/hoogle/).

## The search works roughly like that:

- Fetch the summary list of all available packages.
- Fetch the `documentation.json` of the latest version for each up-to-date package.
- Parse the type annotations of each exposed module.
- Put the type annotations into one big list, also keep a normalised* version with it.
- Parse the search query like a type annotation, calculate a `distance` between the normalised versions.
- Exclude annotations with a `distance` above a given threshold.
- Show search results, sorted by distance.

Normalisation means to replace the type variables with ones in alphabetical order, respecting identical occurrences.

## Please note:

- To see the search in action one has to build the project from source with a set of packages in the `/packages` directory. Packages can be found at https://github.com/elm-lang/package.elm-lang.org/tree/diffs/tmp
- Visiting the search page loads the documentation of all current packages. As the time of writing this is around 165 request, with a total of maybe 1MB. A single bundle file on the server with all documentations would help, but also some client side storage. I'm happy to implement some client side caching, but I'm not sure about what people think about using external dependencies for this website.
- Only up-do-date packages are loaded. It's easy to include all packages, some might not get parsed, though. Any opinions about this?
- Only `Value` annotations are searched, no `Union` or `Alias`. It's easy to add those as well, but is there any value in it?
- The search works either for just the name or the actual annotation. The name-comparison (in both) is super simple and only cares about actual equality or substring occurrences. Hooking into some fuzzy text search would be easy, but I wasn't sure again about bundling a dependency.
- There is some code duplication for the search result type views. I wanted have correct links from there to the corresponding packages, both in for the name and the annotation. But the current views depend on the context (URL) of the page they are in, the search results need to produce absolute links. I did not want to put too much effort into this at this stage. Any help there is appreciated.
- The [`distance` implementation](https://github.com/klaftertief/package.elm-lang.org/blob/feature/type-search/src/frontend/Docs/Type.elm#L237-L277) of two types is not the nicest at the moment. I could use some more experienced eyes there, mainly to find some good penalty settings to have a good constant threshold for filtering the search results. I was thinking about normalising the `distance` by using some complexity measure of each type.
- Searching with type variables (lowercase annotations) is limited such that one can't use a single type variable like `x` to match anything "structured", like `List a`. That is a search for `x -> y -> x` only finds `always` and not some arbitrary function whose return type maths the type of the first parameter.

So, before merging I'd like to have a better notation of `distance` values with the help of a good set of penalties and probably some DRYed up views.

elm-search-empty
elm-search-searching

### CODE REVIEW

File: src/frontend/Component/Search.elm [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+        ]
+    else
+      div [] (searchResultsChunks info filteredChunks)
+
+
+searchResultsChunks : Info -> List ( Int, Chunk ) -> List Html
+searchResultsChunks { packageDict, focusedPackage } weightedChunks =
+  weightedChunks
+    |> List.sortBy (\( distance, _ ) -> distance)
+    --|> List.filter (\( _, { package } ) -> focusedPackage == Nothing || focusedPackage == Just package)
```
Comment (eeue56) [inline src/frontend/Component/Search.elm]: could probably kill this?

File: src/frontend/Docs/Entry.elm [hunk 2]
```diff
@@ -83,6 +83,34 @@ tagMap func tag =
...
+    Value tipe _ ->
+      if query == model.name then
+        0
+      else if String.contains query model.name then
+        1
+      else
+        10
+
+    _ ->
+      100
```
Comment (eeue56) [inline src/frontend/Docs/Entry.elm]: How about sticking these in a constant somewhere so they're easy to change?

File: src/frontend/Docs/Entry.elm [hunk 3]
```diff
@@ -191,6 +269,19 @@ longFunctionAnnotation nameDict args result =
...
 
+longFunctionAnnotationSearch : String -> Name.Dictionary -> List Type -> Type -> List (List Html)
+longFunctionAnnotationSearch basePath nameDict args result =
+  let
+    tipeHtml =
+      List.map (Type.toHtmlWithBasePath basePath nameDict Type.Func) (args ++ [result])
+
+    starters =
+      [ text "    ", colon, text "  " ]
+      ::
```
Comment (eeue56) [inline src/frontend/Docs/Entry.elm]: This reads a bit weirdly

File: src/frontend/Docs/Type.elm [hunk 4]
```diff
@@ -164,3 +229,153 @@ length context tipe =
         in
           recordLength + extLength
 
+
+
+-- SEARCH
+
+
+distance : Type -> Type -> Int
+distance needle hay =
```
Comment (eeue56) [inline src/frontend/Docs/Type.elm]: This function could do with some kind of documentation

File: src/frontend/Docs/Entry.elm (hunk 2 above)
Comment (klaftertief) [inline src/frontend/Docs/Entry.elm]: Yep. Those have to match the ones in `Docs.Type`, missing there as constants as well at the moment. I'm not really sure how many different values we need and how to name them. Needs some exploration.

File: src/frontend/Docs/Entry.elm (hunk 3 above)
Comment (klaftertief) [inline src/frontend/Docs/Entry.elm]: That is the same as in `longFunctionAnnotation`, which is existing code. As said in the PR, this needs to get DRYed up. I'm just not sure if I want to add a `basePath` or `Page.Context` to the existing "normal" views.

**Comment by klaftertief** [comment]
I've updated the code a bit. Still need to prepare for the penalty/threshold constants.

**Comment by evancz** [comment]
So I am not a big fan of Hoogle or Hayoo. I have been using Haskell for something like 6 years now, and it never felt like the easiest way to find the information I am looking for in practice. The one time I remember using it was when I was reading code with `<|>` and tons of unqualified imports such that I couldn't tell where it might be from. My search did not end up working very well even, I found it some other way! Point is, if I am going to accept the maintenance burden of something like this, it makes sense to quantify who this helps and by how much.

So can we outline a few concrete scenarios where folks will want to use this tool, and this will be the most efficient path to the result? Who are these people? What are they trying to build? What is the exact sequence of events they go through?

**Comment by klaftertief** [comment]
I don't have a Haskell background but used Hoogle a bit when I started to play around with Elm to find a function name for some type annotations. So I knew what I needed to have, but I did not know how this thing is called of where to find it. Hoogle helped to get the name, but I stilled had to go and search the `core` and some `-extra` packages to find an implementation.
On the other hand, when you start to know the names of some fundamental functional functions (map, fold, ...) you can now see with what types and packages you can apply your gained knowledge.

So my main scenario is to have a tool, mainly for beginners, to help to get a feeling for functional concepts and to explore the Elm ecosystem.

> If this tool is going to make it easier to write code that breaks these guidelines, I'm not sure it's such a good idea.

I don't know if I understand your concern correctly. Do you mean that one might use this search when she is using a poorly implemented package (with regard to the guidelines) and then people are not motivated to follow the guidelines?

**Comment by eeue56** [comment]
I use Hoogle pretty much consistently whenever I'm doing Haskell work. There's also been many times when I've tried to remember the name of a function in Elm failed, but I know the signature. For example, I might know `a -> List a -> Bool`, but I might not know it's called `List.member`. That's a real example from a day ago when I'd been using `String.contains` too much. Another case is when I'm looking for a function, but I'm not sure what package it's in. For example, let's consider that I want a function like `Int -> List a -> List (List a)` which will take a size, a list, then return the list broken down into a list of lists of the given size. This exists in `elm-list-split`, but without searching on the type signature I never would've found it, nor even know the name of it.

**Comment by klaftertief** [comment]
I’ve put together a demo of the search so that people can evaluate it. It's available at http://klaftertief.github.io/package.elm-lang.org/
It uses a remote CORS proxy, but it also stores the package docs locally to avoid too many requests.

**Comment by gyzerok** [comment]
About 2 days ago I wanted to find something like `Time -> a -> Effects a`. And I thought about types search.

**Comment by achernyak** [comment]
@evancz I see it as a good tool for people starting out with FP or looking for very specific type signatures. This might add to package discoverability also. 

On bigger projects, you don't really see the full benefit of a tool like Hoogle until you have a local instance of it running. This lets you search everything in your project and all of its dependencies.

**Comment by G4BB3R** [comment]
@evancz 
I came from Haskell, and I really missed a Hoogle like tool for Elm.
A lot of times I don't remember the exact function name, or sometimes when I know the type signature but I forget the name, this would be easier than searching in every doc.
It is also useful to code in an idiomatic way, without needing to rewrite some small functions that already exist in the core (or even 3rd party) libraries but you didn't know..
Maybe it will not help too much experienced elm programmers, because if you use every day, you will memorize the most important functions and sometimes look at the docs, but It will really help beginners or intermediate elm programmers to avoid searching manually every momment, functions in the docs.

@klaftertief
In my opinion core libraries should have priority in the results, just a suggestion.

**Comment by lorenzo** [comment]
I would have helped me a lot in the beginning. It took me a while to understand where the operators were documented, and also what "html" package I was using, since for some reason I thought it was part of the core. so :+1:

**Comment by reactormonk** [comment]
https://klaftertief.github.io/elm-search/ works quite well - what happened to this PR?

**Comment by klaftertief** [comment]
I wanted to wait until I officially announced elm-search to close this pull request. But as it's online since a while I guess There is no need to wait.

**Comment by klaftertief** [comment]
I'll open a separate pull request with a link to the search to (package.)elm-lang.org then.

**Pull request closed without merging**